hbase_sql/src/main/java/com/splicemachine/derby/stream/function/HFileGenerationFunction.java

```diff
@@ -0,0 +1,214 @@
+import org.apache.spark.sql.Row;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Created by jyuan on 3/21/17.
+ */
+public class HFileGenerationFunction<U> implements MapPartitionsFunction<Row, U>, Externalizable {
+
+    private static final Logger LOG=Logger.getLogger(HFileGenerationFunction.class);
+
+    private List<U> HFiles = new ArrayList<U>();
+    private boolean initialized;
+    private FileSystem fs;
+    private long txnId;
+    private Configuration conf;
+    private OperationContext operationContext;
+    private Long heapConglom;
+    private List<BulkImportPartition> partitionList;
+
+    private Long currentCongomerate;
+    private BulkImportPartition currentPartition;
+    private StoreFile.Writer writer;
+
+    public HFileGenerationFunction() {
+    }
+
+    public HFileGenerationFunction(OperationContext operationContext,
+                                   long txnId,
+                                   Long heapConglom,
+                                   List<BulkImportPartition> partitionList) {
+        this.txnId = txnId;
+        this.operationContext = operationContext;
+        this.heapConglom = heapConglom;
+        this.partitionList = partitionList;
+    }
+
+    public Iterator<U> call(Iterator<Row> mainAndIndexRows) throws Exception {
+
+        while(mainAndIndexRows.hasNext()) {
+            Row row = mainAndIndexRows.next();
+            Long conglomerateId = row.getAs("conglomerateId");
+            byte[] key = Bytes.fromHex(row.getAs("key"));
+            byte[] value = row.getAs("value");
+            if (LOG.isDebugEnabled()) {
+                SpliceLogUtils.error(LOG, "conglomerateId:%d, key:%s, value:%s",
+                        conglomerateId, Bytes.toHex(key), Bytes.toHex(value));
+            }
+            if (!initialized) {
+                init(conglomerateId, key);
+                initialized = true;
+            }
+            if (!belongToCurrentPartititon(conglomerateId, key)) {
+                close(writer);
+                prepareNextFile(conglomerateId, key);
+            }
+            writeToHFile(key, value);
+            if (conglomerateId.equals(heapConglom)) {
+                operationContext.recordWrite();
+            }
+        }
+        close(writer);
+        if (HFiles ==null || HFiles.size() == 0)  {
+            HFiles.add((U)"Empty");
+        }
+
+        return HFiles.iterator();
+    }
+
+    private void writeToHFile (byte[] rowKey, byte[] value) throws Exception {
+        KeyValue kv = new KeyValue(rowKey, SIConstants.DEFAULT_FAMILY_BYTES,
+                SIConstants.PACKED_COLUMN_BYTES, txnId, value);
+        writer.append(kv);
+
+    }
+
+    private boolean belongToCurrentPartititon(Long conglomerateId, byte[] key) {
```

Comment:
Is there a way to get rid of this? They typical case is that it is the same partition, but we are paying the penalty of the comparison over and over